libraries/AP_Scripting/examples/quadruped.lua

```diff
@@ -0,0 +1,308 @@
+
+-- starting positions of the legs
+local endpoints1 = {math.cos(45/180*math.pi)*(COXA_LEN + FEMUR_LEN), math.sin(45/180*math.pi)*(COXA_LEN + FEMUR_LEN), TIBIA_LEN }
+local endpoints2 = {math.cos(45/180*math.pi)*(COXA_LEN + FEMUR_LEN), math.sin(-45/180*math.pi)*(COXA_LEN + FEMUR_LEN), TIBIA_LEN }
+local endpoints3 = {-math.cos(45/180*math.pi)*(COXA_LEN + FEMUR_LEN), math.sin(-45/180*math.pi)*(COXA_LEN + FEMUR_LEN), TIBIA_LEN }
+local endpoints4 = {-math.cos(45/180*math.pi)*(COXA_LEN + FEMUR_LEN), math.sin(45/180*math.pi)*(COXA_LEN + FEMUR_LEN), TIBIA_LEN }
+
+-- control input enum
+local control_input_roll = 1
+local control_input_pitch = 2
+local control_input_throttle = 3
+local control_input_yaw = 4
+
+local xy_speed_max = 40     -- x and y axis speed max (used to convert control input) in mm
+local yaw_speed_max = 10    -- yaw speed maximum (used to convert control input)
+local speed_dz = 5          -- speed deadzone.  x, y and yaw speed requests are ignored if their absolute value is less than this number
+local x_speed = 0           -- forward speed target
+local y_speed = 0           -- lateral speed target (right is positive, left is negative)
+local yaw_speed = 0         -- yaw rotation speed target
+
+local leg_lift_height = 50  -- leg lift height (in mm) while walking
+
+-- gait definition parameters
+local gait_type = 0         -- gait pattern.  0 = alternating gait, 1 = wave gait.
+local gait_step = 0         -- gait step in execution
+local gait_step_total = 0   -- number of steps in gait
+local gait_step_leg_start = {0,0,0,0} -- leg starts moving on this gait step (front-right, front-left, back-left, back-right)
+local gait_lifted_steps = 0 -- number of steps that a leg is lifted for
+local gait_down_steps = 0   -- number of steps that the leg lifted needs to be put down for
+local gait_lift_divisor = 0 -- when a leg is lifted and brought back down the action is divided into 2 or multiple steps, so the travel distance also need to be split in between the steps to make the transition natural
+local gait_half_lift_height = 0 -- used to split lift across two steps
+local gait_speed_divisor = 0    -- number of steps in the gait the leg is touching the floor, this is used as a factor to split the travel distance between the steps
+
+local gait_pos_x = {0,0,0,0}    -- X-axis position for each leg (front-right, front-left, back-right, back-left)
+local gait_pos_y = {0,0,0,0}    -- Y-axis position for each leg (front-right, front-left, back-right, back-left)
+local gait_pos_z = {0,0,0,0}    -- Z-axis position for each leg (front-right, front-left, back-right, back-left)
+local gait_rot_z = {0,0,0,0}    -- Z-axis rotation for each leg (front-right, front-left, back-right, back-left)
+local leg_index = 0
+local last_angle = {0,0,0,0,0,0,0,0,0,0,0,0}
+local current = {0,0,0,0,0,0,0,0,0,0,0,0}
+local start_time = 0
+local curr_target = 0
+local pwm = {1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500}
```

Comment:
This whole array appears to only be needed in one function, where the PWM is calculated then output, why not remove this, and directly output it? It would reduce the memory cost here, and should be faster.

Reply:
txs, moved pwm to be local to the update function

Reply:
Why even have it local? You compute the value, store it in the array, then immediately send it to an output. Why not send the value to the output without ever having the array to store it in? (Saves the entire array from being in memory)

Reply:
yes, fair enough, zapped!